### PR TITLE
(486) Refactoring Workflow

### PR DESCRIPTION
**Pull request by kekonen**
#486 - Refactor Workflow
https://github.com/OpusCapita/businessnetwork/issues/133 - email events
https://github.com/OpusCapita/businessnetwork/issues/122 - event driven

---

Refactored workflow.js a lot:
- changed endpoints to events
- deleted unused functions and endpoints (together with ACL rules)
- moved one left endpoint 'startCampaign' to campaignContact.js 
- ! left emailOpened event based function in a test state, since it isnot possible to check it on local (white external address is needed)

### CODE REVIEW

File: src/server/workflow/workflow.js [hunk 1]
```diff
@@ -12,602 +10,420 @@ const Templates = require('../api/Templates');
...
+				model: db.models.Campaign,
+				required: true
+			},
+			where: {
+				id: contactInfo.id
+			}
+		})
+		.then(contact => {
+			let data = {
+				type: 'singleUse',
```
Comment (gr4per) [inline src/server/workflow/workflow.js]: some comment, explaining why we are in the targeted campaign case here always

File: src/server/workflow/workflow.js [hunk 2]
```diff
@@ -12,602 +10,420 @@ const Templates = require('../api/Templates');
...
+
+	                        return client.post('email', '/api/send', {
+	                                to: contact.email,
+	                                subject,
+	                                html: result
+	                            }, true)
+	                            .then(transactionId => {
+	                                if (transactionId) {
+	                                    return contact.update({
+	                                        messageId: transactionId[0][0].messageId,
```
Comment (gr4per) [inline src/server/workflow/workflow.js]: looks ugly. why dont we have a proper property access path ?

File: src/server/workflow/workflow.js [hunk 3]
```diff
@@ -12,602 +10,420 @@ const Templates = require('../api/Templates');
...
+	                                        status: 'sent',
+	                                        lastStatusChange: new Date()
+	                                    })
+
+	                                    logger.info('Email sent. (contactId: %s)', contact.id);
+	                                }
+	                            })
+	                            .catch(e => {
+	                                return db.models.CampaignContact.update({
+	                                    status: 'bounced',
```
Comment (gr4per) [inline src/server/workflow/workflow.js]: This is wrong, if the api call to email fails it wasnt even sent, so how can it bounce?

Bounce is an explicit event we get by emails service only after successful sending

File: src/server/workflow/workflow.js [hunk 4]
```diff
@@ -12,602 +10,420 @@ const Templates = require('../api/Templates');
...
+    };
+
+
+
+	this.eventClient.subscribe('user.onboardingdata.created', processOnboardCreation);
+
+	this.eventClient.subscribe('onboarding.campaignContact.workflow.start', launchContact);
+	this.eventClient.subscribe('onboarding.campaignContact.updated.invitationGenerated', sendMail);
+
+	this.eventClient.subscribe('onboarding.*.opened', emailOpened);
```
Comment (gr4per) [inline src/server/workflow/workflow.js]: This needs change. Star needs replacement by customEmailTopic, so i guess pattern would be
email.onboarding.opened

but we need to listen on "all" events by mailjet related to this custom topic, and set contact status accordingly:

opened
bounced
spam
...
check mailjet/email docs

File: src/server/acl.json [hunk 5]
```diff
@@ -1,126 +1,118 @@
...
-    "resources": [{
-        "type": "rest",
-        "resourceId": "/api/getWorkflowTypes",
-        "actions": ["view"]
-      },
+    "resources": [
       {
         "type": "rest",
-        "resourceId": "/api/campaigns/start/[%w | %p]*",
+        "resourceId": "^/api/campaigns/start/[^/]+/[^/]+",
```
[the remark this answers is not in the document]
Comment (kekonen) [reply src/server/acl.json]: The endpoints which end with  `/[^/]+` are secure enough and don't need to have $ in the end. Ref: Chris

File: src/server/workflow/workflow.js [hunk 6]
```diff
@@ -12,475 +10,56 @@ const Templates = require('../api/Templates');
...
+    });
 
-  function processUserUpdated(userData) {
-    console.log('processing user.updated event, user ' + userData.id + ", status " + userData.status);
-    if(userData.status === 'emailVerified') {
-      this.client.get('user', '/onboardingdata/'+userData.id, true).spread((onboardData, response) => {
-        if (onboardData
-          && onboardData.campaignTool === 'opuscapitaonboarding'
-          && onboardData.invitationCode) {
+    this.campaignToolName = CAMPAIGNTOOLNAME;
```
[the remark this answers is not in the document]
Comment (kekonen) [reply src/server/workflow/workflow.js]: It was there before, just changed place in order to keep the file clean. Basically, name of the workflow.

File: src/server/db/models/Campaign.js [hunk 7]
```diff
@@ -162,6 +163,11 @@ module.exports.init = function(sequelize)
     {
       type: Sequelize.STRING(30),
       allowNull: true
+    },
+	emailMessageId:
```
[the remark this answers is not in the document]
Comment (nczeranka) [reply src/server/db/models/Campaign.js]: It is used now for tracking the email.

File: src/server/db/migrations/201804131500000-workflowStatusNames.main.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        ...(queued.length>0?[db.queryInterface.sequelize.query("UPDATE CampaignContact SET `status`='queued' WHERE `status`='errorGeneratingInvitation' AND `id` IN(:ids)", { replacements: { ids: queued }, type: db.queryInterface.sequelize.QueryTypes.BULKUPDATE })]:[]),
+        ...(generatingInvitation.length>0?[db.queryInterface.sequelize.query("UPDATE CampaignContact SET `status`='generatingInvitation' WHERE `status`='errorGeneratingInvitation' AND `id` IN(:ids)", { replacements: { ids: generatingInvitation }, type: db.queryInterface.sequelize.QueryTypes.BULKUPDATE })]:[]),
+        ...(needsVoucher.length>0?[db.queryInterface.sequelize.query("UPDATE CampaignContact SET `status`='needsVoucher' WHERE `status`='errorServiceConfig' AND `id` IN(:ids)", { replacements: { ids: needsVoucher }, type: db.queryInterface.sequelize.QueryTypes.BULKUPDATE })]:[]),
+        ...(generatingVoucher.length>0?[db.queryInterface.sequelize.query("UPDATE CampaignContact SET `status`='generatingVoucher' WHERE `status`='errorServiceConfig' AND `id` IN(:ids)", { replacements: { ids: generatingVoucher }, type: db.queryInterface.sequelize.QueryTypes.BULKUPDATE })]:[]),
+    ]
+
+    const toDowSecond = [
```
Comment (gr4per) [inline src/server/db/migrations/201804131500000-workflowStatusNames.main.js]: toDownSecond

File: src/server/routes/campaign.js [hunk 9]
```diff
@@ -242,8 +244,65 @@ module.exports = (app, db) => {
...
+    			},
+    			where: {
+    				status: 'new'
+                    // {
+                    //     $in: ['new', 'failedToSend']
+                    // }
+    			}
+    		})
+            .map(contact => {
+                return this.eventClient.emit('onboarding.campaignContact.workflow.start', contact)
```
Comment (gr4per) [inline src/server/routes/campaign.js]: event-client doesn't know your current request scope hence it won't be able to inject any context info (correlationId, userId etc.)

Comment (kwierchris) [reply src/server/routes/campaign.js]: I think you can change **this.eventClient** with **req.opuscapita.eventClient** here and remove it's initialization.

File: src/client/containers/CampaignOverview.react.js [hunk 10]
```diff
@@ -90,7 +90,7 @@ class CampaignOverview extends Components.ConditionalRenderComponent
         return new Promise((resolve, reject) =>
         {
             const { i18n, showNotification, router } = this.context;
-            const { campaignId, emailSendCount } = this.state;
+            const { campaignId, emailSendCount, customerId } = this.state;
```
[the remark this answers is not in the document]
Comment (kekonen) [reply src/client/containers/CampaignOverview.react.js]: Below, this `customerId` is used to create an actual `tenantId`, but before that it, there is no `tenantId` in the context

[the remark this answers is not in the document]
Comment (gr4per) [reply src/client/containers/CampaignOverview.react.js]: tenantId is part of the campaign unique key. There is no way we dont have it. Create a new ticket for refactoring the customer crap to tenant everywhere. We will have supplier campaigns, using customerId is just totally wrong.

File: src/client/containers/CampaignOverview.react.js [hunk 11]
```diff
@@ -109,7 +109,7 @@ class CampaignOverview extends Components.ConditionalRenderComponent
             {
                 if(button === 'yes')
                 {
-                    return this.campaignsApi.startCampaign(this.state.campaignId).then(() =>
+                    return this.campaignsApi.startCampaign(campaignId, 'c_' + customerId).then(() =>
                     {
```
Comment (gr4per) [inline src/client/containers/CampaignOverview.react.js]: This is not the correct way. We have util functions in bouncer that allow to convert tenatn ids to customerid / supplierId and the other way round.

File: src/server/acl.json [hunk 12]
```diff
@@ -1,108 +1,112 @@
...
-        "type": "rest",
-        "resourceId": "/api/getWorkflowTypes",
-        "actions": ["view"]
-      },
+    "resources": [
       {
         "type": "rest",
-        "resourceId": "/api/campaigns/start/[%w | %p]*",
+        "resourceId": "^/api/campaigns/start/${_current_tenant_id}/[^/]+$",
         "actions": ["edit"]
```
Comment (gr4per) [inline src/server/acl.json]: we will need to change this for multi tenant access, but ok for now.

File: src/server/acl.json [hunk 13]
```diff
@@ -1,108 +1,112 @@
...
       {
         "type": "rest",
-        "resourceId": "/api/templates/${_current_customer_id}",
+        "resourceId": "^/api/templates/${_current_customer_id}$",
         "actions": ["view", "create"]
       },
       {
         "type": "rest",
-        "resourceId": "/api/templates/${_current_customer_id}/[%w | %p]*",
+        "resourceId": "^/api/templates/${_current_customer_id}/[^/]+$",
```
Comment (gr4per) [inline src/server/acl.json]: This also wont survie the next day. It need to be tenantId

File: src/server/acl.json [hunk 14]
```diff
@@ -1,108 +1,112 @@
...
+        "actions": ["view", "create", "edit", "delete"]
+      },
+      {
+        "type": "rest",
+        "resourceId": "^/api/templates/${_current_customer_id}/[^/]+/[^/]+$",
         "actions": ["view", "create", "edit", "delete"]
```
Comment (gr4per) [inline src/server/acl.json]: same here

File: src/server/api/TemplateError.js [hunk 15]
```diff
@@ -7,6 +13,41 @@ class TemplateError extends Error
...
+
+        if (!flags) flags = {
+            en: 'uk',
+            de: 'de'
+        };
+
+        if (!languageLiterals) languageLiterals = {
+            en: 'English',
+            de: 'Deutch'
+        };
```
Comment (gr4per) [inline src/server/api/TemplateError.js]: Rejected. Translations have to go into json bundles in the i18n folder

File: src/server/api/TemplateError.js [hunk 16]
```diff
@@ -7,6 +13,41 @@ class TemplateError extends Error
...
+
+        if (!contactData) contactData = {
+            en: 'In case of questions, please contact our support at +49 231 3967 350 or via email at ',
+            de: 'Bei Fragen kontaktieren Sie bitte unseren Support unter +49 231 3967 350 oder via E-Mail.'
+        }
```
Comment (gr4per) [inline src/server/api/TemplateError.js]: Rejected. Translations have to go into json bundles in the i18n folder

File: src/server/routes/templates.js [hunk 17]
```diff
@@ -72,107 +76,126 @@ TemplateWebApi.prototype.renderTemplate = function(req, res)
...
+                    translations: this.i18n.getMessageObject('template.error.linkNotSupported'),
+                    language: 'de'
+                });
+            })
+        } else {
+            // Delete this whole 'else' block, when we stop accepting /public/(landingpage/email)/:tenantId/:campaignId/:contactId endpoints
+            // new 'else' would be a fallback
+            var {tenantId, campaignId, contactId} = req.params;
+            var customerId = tenantId.startsWith('c_') ? tenantId.substring(2) : tenantId;
+            invitationCode = req.query.invitationCode;
```
Comment (gr4per) [inline src/server/routes/templates.js]: Not ok. Need to use common util methods

**Comment by kekonen** [comment]
The issues with customerId/tenantId must be solved in future in scope of this issue https://github.com/OpusCapita/useridentity-middleware/issues/10

File: src/server/acl.json (hunk 12 above)
Comment (kekonen) [reply src/server/acl.json]: In scope of https://github.com/OpusCapita/onboarding/issues/498

File: src/client/containers/CampaignOverview.react.js (hunk 11 above)
Comment (kekonen) [reply src/client/containers/CampaignOverview.react.js]: In scope of #498 and https://github.com/OpusCapita/useridentity-middleware/issues/10

File: src/server/acl.json (hunk 13 above)
Comment (kekonen) [reply src/server/acl.json]: In scope of #498

File: src/server/acl.json (hunk 14 above)
Comment (kekonen) [reply src/server/acl.json]: In scope of #498

File: src/server/api/TemplateError.js (hunk 16 above)
Comment (kekonen) [reply src/server/api/TemplateError.js]: done

File: src/server/routes/templates.js (hunk 17 above)
Comment (kekonen) [reply src/server/routes/templates.js]: done

**Pull request merged**